jsonrpc/debug_endpoint.go

```diff
@@ -0,0 +1,209 @@
+package jsonrpc
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/0xPolygon/polygon-edge/helper/hex"
+	"github.com/0xPolygon/polygon-edge/state"
+	"github.com/0xPolygon/polygon-edge/state/runtime/tracer"
+	"github.com/0xPolygon/polygon-edge/state/runtime/tracer/structtracer"
+	"github.com/0xPolygon/polygon-edge/types"
+)
+
+type debugBlockchainStore interface {
+	// Header returns the current header of the chain (genesis if empty)
+	Header() *types.Header
+
+	// GetHeaderByNumber gets a header using the provided number
+	GetHeaderByNumber(uint64) (*types.Header, bool)
+
+	// ReadTxLookup returns a block hash in which a given txn was mined
+	ReadTxLookup(txnHash types.Hash) (types.Hash, bool)
+
+	// GetBlockByHash gets a block using the provided hash
+	GetBlockByHash(hash types.Hash, full bool) (*types.Block, bool)
+
+	// GetBlockByNumber gets a block using the provided height
+	GetBlockByNumber(num uint64, full bool) (*types.Block, bool)
+
+	TraceMinedBlock(*types.Block, tracer.Tracer) ([]interface{}, error)
+
+	TraceMinedTxn(*types.Block, types.Hash, tracer.Tracer) (interface{}, error)
+
+	TraceCall(*types.Transaction, *types.Header, tracer.Tracer) (interface{}, error)
+}
+
+type debugTxPoolStore interface {
+	GetNonce(types.Address) uint64
+}
+
+type debugStateStore interface {
+	GetAccount(root types.Hash, addr types.Address) (*state.Account, error)
+}
+
+type debugStore interface {
+	debugBlockchainStore
+	debugTxPoolStore
+	debugStateStore
+}
+
+// Debug is the debug jsonrpc endpoint
+type Debug struct {
+	store debugStore
+}
+
+type TraceConfig struct {
+	EnableMemory     bool `json:"enableMemory"`
+	DisableStack     bool `json:"disableStack"`
+	DisableStorage   bool `json:"disableStorage"`
+	EnableReturnData bool `json:"enableReturnData"`
+	// Timeout *string `json:"timeout"`
+}
+
+func (d *Debug) TraceBlockByNumber(
+	blockNumber BlockNumber,
+	config *TraceConfig,
+) (interface{}, error) {
+	num, err := GetNumericBlockNumber(blockNumber, d.store)
+	if err != nil {
+		return nil, err
+	}
+
+	block, ok := d.store.GetBlockByNumber(num, true)
+	if !ok {
+		return nil, errors.New("block not found")
```

Comment:
could we always include user's context to errors? like block number or hash?

Reply:
as a separate task, how do you think, do we need error rate per API or something like that in our metrics?

Reply:
Added 39fdd3601694f705f6bd563b89848aee94273fe1